### PR TITLE
removed the calculation in the noteblock block entity

### PR DESCRIPTION
**Pull request by 12xx12**
I looked for todo's in the code and this is the first one I know how to fix

I did the calculation in python if anyone is interested where the numbers are from
```python 
for i in range(0,25):
    print(2**((i-12)/12))
```

### CODE REVIEW

File: src/BlockEntities/NoteEntity.cpp [hunk 1]
```diff
@@ -245,8 +245,35 @@ void cNoteEntity::MakeSound(void)
...
 	m_World->BroadcastBlockAction(m_Pos, static_cast<Byte>(instrument), static_cast<Byte>(m_Pitch), E_BLOCK_NOTE_BLOCK);
 
-	// TODO: instead of calculating the power function over and over, make a precalculated table - there's only 24 pitches after all
-	float calcPitch = static_cast<float>(pow(2.0f, static_cast<float>(m_Pitch - 12.0f) / 12.0f));
+	float calcPitch;
+	switch (m_Pitch)
+	{
+		case 0: calcPitch = 0.5f;
+		case 1: calcPitch = 0.5297315471796477f;
+		case 2: calcPitch = 0.5612310241546865f;
```
Comment (tigerw) [inline src/BlockEntities/NoteEntity.cpp:253]: I couldn't bother you to make this into a static function, could I? As this code stands the lack of `break` means execution falls all the way to the bottom and `calcPitch` ends up always with value `2`.

Comment (12xx12) [reply src/BlockEntities/NoteEntity.cpp:253]: You find my lack of breaks disturbing...

I could move this to a static function of course

Comment (12xx12) [reply src/BlockEntities/NoteEntity.cpp]: i hope the new commit is in the way you asked for

**Comment by 12xx12** [comment]
I think I'm gonna change the m_Pitch to m_Note to prevent confusion

**Comment by Bond-009** [comment]
I’d add the original formula as comment somewhere in CalculatePitch.
Also, CalculatePitch is a misnomer as it doesn’t really calculate anything

File: src/BlockEntities/NoteEntity.h [hunk 2]
```diff
@@ -56,6 +56,10 @@ class cNoteEntity :
 	virtual bool UsedBy(cPlayer * a_Player) override;
 	virtual void SendTo(cClientHandle &) override {}
 
+	/** Calculates the pitch to be send to the players
+	from a pitch value from 0 to 24 as incremented by the IncrementPitch function */
```
Comment (madmaxoft) [inline src/BlockEntities/NoteEntity.h:60]: How about:
```suggestion
	/** Returns the relative pitch (used in the protocol)
	from a note value between 0 and 24 (used in m_Pitch). */
```

File: src/BlockEntities/NoteEntity.h [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 
 
-enum ENUM_NOTE_INSTRUMENTS
+enum ENUm_Pitch_INSTRUMENTS
```
Comment (madmaxoft) [inline src/BlockEntities/NoteEntity.h:10]: Why rename this in such a weird fashion?

File: src/BlockEntities/NoteEntity.h [hunk 4]
```diff
@@ -56,6 +56,10 @@ class cNoteEntity :
 	virtual bool UsedBy(cPlayer * a_Player) override;
 	virtual void SendTo(cClientHandle &) override {}
 
+	/** Calculates the pitch to be send to the players
+	from a pitch value from 0 to 24 as incremented by the IncrementPitch function */
+	static float GetPitch(char a_Note);
```
Comment (madmaxoft) [inline src/BlockEntities/NoteEntity.h:61]: Perhaps also rename to `PitchFromNote` to be absolutely clear about this?

**Review by madmaxoft — CHANGES_REQUESTED** [review]
I quite liked the renaming of `m_Pitch` to `m_Note`, makes much more sense. Changing the API is okay in this case, you'll just need to provide a compatibility stub in src/Bindings/DeprecatedBindings.cpp (there are a lot of examples there), so that plugins using the old function names will still work, but emit a warning to the console.

**Comment by madmaxoft** [comment]
Let me know if you need help with the deprecated bindings.

File: src/BlockEntities/NoteEntity.h (hunk 2 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h:60]: ```(used in m_Note)``` I think

Comment (madmaxoft) [reply src/BlockEntities/NoteEntity.h:60]: Yes, if you rename it again :)

File: src/BlockEntities/NoteEntity.h (hunk 4 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h:61]: Thats a nice name

File: src/BlockEntities/NoteEntity.h (hunk 3 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h:10]: This is a good Question - didn't notice that.. used the rename tool and this was the result - gonna change that

File: src/BlockEntities/NoteEntity.h (hunk 4 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h]: Renamed

**Comment by 12xx12** [comment]
I thing I got the binding

File: src/BlockEntities/NoteEntity.h (hunk 3 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h]: reverted changes

File: src/BlockEntities/NoteEntity.h (hunk 2 above)
Comment (12xx12) [reply src/BlockEntities/NoteEntity.h]: changed function documentation

**Comment by 12xx12** [comment]
ok - i need help with the deprecated bindings

**Comment by tigerw** [comment]
For what it's worth I'm not seeing any calls to DoWithNoteBlockAt in Lua files when doing a GitHub code search. Probably very few plugins manipulate noteblocks.

File: src/Bindings/DeprecatedBindings.cpp [hunk 5]
```diff
@@ -599,6 +599,13 @@ void DeprecatedBindings::Bind(lua_State * tolua_S)
 		tolua_function(tolua_S, "UpdateSign", tolua_cWorld_SetSignLines);
 	tolua_endmodule(tolua_S);
 
+
+	tolua_beginmodule(tolua_S, "cNoteEntity");
+		tolua_function(tolua_S, "GetPitch", tolua_AllToLua_cNoteEntity_GetNote00);
+		tolua_function(tolua_S, "SetPitch", tolua_AllToLua_cNoteEntity_SetNote00);
+		tolua_function(tolua_S, "IncrementNote", tolua_AllToLua_cNoteEntity_IncrementNote00);
```
Comment (madmaxoft) [inline src/Bindings/DeprecatedBindings.cpp:606]: You can't "redirect" to a handler in automatic bindings - they are marked `static` so they are unavailable from other source files; also we want to output a warning and stacktrace to the console if these old functions are used.

Instead, create new functions for these that remaps the call:
```cpp
static int tolua_cNoteEntity_GetPitch(lua_State * tolua_S)
{
	cLuaState LuaState(tolua_S);

	if (
		!LuaState.CheckParamUserType(1, "cNoteEntity") ||
		!LuaState.CheckParamEnd(2)
	)
	{
		return 0;
	}

	cNoteEntity * Self = nullptr;

	if (!LuaState.GetStackValues(1, Self))
	{
		tolua_error(LuaState, "Failed to read parameters", nullptr);
	}
	if (Self == nullptr)
	{
		tolua_error(LuaState, "invalid 'self' in function 'GetNote'", nullptr);
	}
	LuaState.Push(Self->GetNote());
	LOGWARNING("Warning: 'cNoteEntity:GetPitch' function is deprecated. Please use 'cNoteEntity:GetNote' instead.");
	LuaState.LogStackTrace(0);
	return 1;
}
```

Also, please alpha-sort both the functions in a module, and module by name.

**Pull request merged**